python/dgl/view.py

```diff
@@ -282,33 +282,75 @@ def __call__(self, ntype=None):
 
 class HeteroNodeDataView(MutableMapping):
     """The data view class when G.ndata[ntype] is called."""
-    __slots__ = ['_graph', '_ntype', '_ntid', '_nodes']
+    __slots__ = ['_graph', '_ntype', '_ntid', '_nodes', '_multi_ntype']
 
     def __init__(self, graph, ntype, ntid, nodes):
         self._graph = graph
         self._ntype = ntype
         self._ntid = ntid
         self._nodes = nodes
+        self._multi_ntype = isinstance(ntype, list)
 
     def __getitem__(self, key):
-        return self._graph._get_n_repr(self._ntid, self._nodes)[key]
+        if self._multi_ntype:
+            ret = {}
+            for (i, ntype) in enumerate(self._ntype):
+                value = self._graph._get_n_repr(self._ntid[i], self._nodes).get(key, None)
+                if value is not None:
+                    ret[ntype] = value
+            return ret
+        else:
+            return self._graph._get_n_repr(self._ntid, self._nodes)[key]
 
     def __setitem__(self, key, val):
-        self._graph._set_n_repr(self._ntid, self._nodes, {key : val})
+        if self._multi_ntype:
+            assert isinstance(val, dict), \
+                'Current HeteroNodeDataView has multiple node types, ' \
+                'please passing the node type and the corresponding data through a dict.'
+
+            for (ntype, data) in val.items():
+                ntid = self._graph.get_ntype_id(ntype)
+                self._graph._set_n_repr(ntid, self._nodes, {key : data})
+        else:
+            assert isinstance(val, dict) is False, \
+                'The HeteroNodeDataView has only one node type. ' \
+                'please pass a tensor directly'
```

Comment:
The check is strange. I think you want to check `val` is of tensor type instead.

Reply:
This assert is to give hint to someone if he pass a dict to a graph with only one edge type.

Reply:
So the following is not allowed?
```python
g = ...  # some homograph with node type 'user'
g.ndata['h'] = {'user': user_feat}
```

Reply:
Yes.
For homograph, only use tensor, otherwise this may be confusion, when there is only one node type, one can use dict and also tensor.